packages/design-system/src/components/tooltip/index.js

```diff
@@ -129,19 +129,35 @@ function Tooltip({
     [placement]
   );
 
-  // When near the bottom of the viewport and the tooltip is placed on the bottom we want to force the tooltip to the top as to not
+  // When near the edge of the viewport we want to force the tooltip to a new placement as to not
   // cutoff the contents of the tooltip.
   const positionPlacement = useCallback(
     ({ offset }) => {
-      // check to see if there's an overlap with the window edge
+      // check to see if there's an overlap with the window's bottom edge
       const neededVerticalSpace = offset.bottom;
-      // if the tooltip was assigned as bottom we want to always check it
-      if (
-        offset &&
+      const shouldMoveToTop =
         dynamicPlacement.startsWith('bottom') &&
-        neededVerticalSpace >= window.innerHeight
-      ) {
+        neededVerticalSpace >= window.innerHeight;
+
+      // in RTL mode, we can sometimes render a tooltip too far to the left
+      const isOverFlowingLeft = offset.popupLeft < 0;
+
+      if (shouldMoveToTop && !isOverFlowingLeft) {
         setDynamicPlacement(PLACEMENT.TOP);
+      } else if (shouldMoveToTop && isOverFlowingLeft) {
+        setDynamicPlacement(PLACEMENT.TOP_START);
+      } else if (!shouldMoveToTop && isOverFlowingLeft) {
+        const currentPlacement = placementRef.current;
+        switch (currentPlacement) {
+          case currentPlacement.endsWith('start'):
+            setDynamicPlacement(currentPlacement.replace('-start', '-end'));
+            break;
+          case currentPlacement.endsWith('end'):
+            setDynamicPlacement(currentPlacement.replace('-end', '-start'));
+            break;
+          default:
+            setDynamicPlacement(`${currentPlacement}-start`);
+        }
```

Comment:
Agreed. It is super weird. At one point I did pull it out, but ended up refactoring it back in, though I had a bit of other stuff going on then. I think I will pull it out and maybe add a little more comments since it is hard to follow.